docs/scalars.rst

```diff
@@ -0,0 +1,176 @@
+            @Override
+            public Object serialize(Object dataFetcherResult) {
+                return serializeEmail(dataFetcherResult);
+            }
+
+            @Override
+            public Object parseValue(Object input) {
+                return parseEmailFromVariable(input);
+            }
+
+            @Override
+            public Object parseLiteral(Object input) {
+                return parseEmailFromAstLiteral(input);
+            }
+        });
+
+
+Coercing values
+---------------
+
+The real work in any custom scalar implementation is the ``graphql.schema.Coercing`` implementation.  This is responsible for 3 functions
+
+* ``parseValue`` - takes a variable input object and converts into the Java runtime representation
+* ``parseLiteral`` - takes an AST literal ``graphql.language.Value` as input and converts into the Java runtime representation
+* ``serialize`` - takes a Java object and converts it into the output shape for that scalar
+
+So your custom scalar code has to handle 2 forms of input (parseValue / parseLiteral)  and 1 form of output (serialize).
+
+Imagine this query, which uses variables, AST literals and outputs our scalar type ```email``.
+
+.. code-block:: graphql
+
+    mutation Contact($mainContact: Email!) {
+      makeContact(mainContactEmail: $mainContact, backupContactEmail: "backup@company.com") {
+        id
+        mainContactEmail
+      }
+    }
+
+Our custom Email scalar will
+
+* be called via ``parseValue`` to convert the ``$mainContact`` variable value into a runtime object
+* be called via ``parseLiteral`` to convert the AST ``graphql.language.StringValue`` "backup@company.com" into a runtime object
+* be called via ``serialise`` to turn the runtime representation of mainContactEmail into a form ready for output
+
+Validation of input and output
+------------------------------
+
+The methods can validate that the received input makes sense.  For example our email scalar will try to validate that the input
+and output are indeed email addresses.
+
+The JavaDoc method contract of ``graphql.schema.Coercing`` says the following
+
+* The ``serialise`` MUST ONLY allow ``graphql.schema.CoercingSerializeException`` to be thrown from it.  This indicates that the
+value cannot be serialised into an appropriate form.  You must not allow other runtime exceptions to escape this method to get
+the normal graphql behaviour for validation.  You MUST return a non null value
+
+
+* The ``parseValue`` MUST ONLY allow ``graphql.schema.CoercingParseValueException`` to be thrown from it.  This indicates that the
+value cannot be parsed as input into an appropriate form.  You must not allow other runtime exceptions to escape this method to get
+the normal graphql behaviour for validation.  You MUST return a non null value.
+
+* The ``parseLiteral``  MUST ONLY allow ``graphql.schema.CoercingParseValueException`` to be thrown from it.  This indicates that the
```

Comment:
I really think we should rename it to `CoercingParseValueOrLiteralException` or something similar: I am happy to use the same exception in order to not make it not too difficult, but `CoercingParseValueException` is a perfect macht for `parseValue` and it is just confusing to use it it for `parseLiteral` too without aligning the name.